### PR TITLE
[TIMOB-24629] Android: Add Notification wakeLock

### PR DESCRIPTION
**Pull request by m1ga**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-24629

Example:
```JS
var win = Ti.UI.createWindow({});
var btn = Ti.UI.createButton({
	title: "click"
});
win.add(btn);

btn.addEventListener("click", function() {
	_.delay(function() {
		var intent = Ti.Android.createIntent({
			action: Ti.Android.ACTION_MAIN,
			className: 'com.appcelerator.notificationsample.NotificationsampleActivity',
			packageName: 'com.appcelerator.notificationsample'
		});
		intent.flags |= Ti.Android.FLAG_ACTIVITY_CLEAR_TOP | Ti.Android.FLAG_ACTIVITY_NEW_TASK;
		intent.addCategory(Ti.Android.CATEGORY_LAUNCHER);

		var pending = Titanium.Android.createPendingIntent({
			intent: intent,
			flags: Titanium.Android.FLAG_UPDATE_CURRENT
		});

		var notification = Titanium.Android.createNotification({
			contentTitle: 'Something Happened',
			contentText: 'Click to return to the application.',
			contentIntent: pending, 
			wakeLock: {
				time: 2000,
				flags: Ti.Android.WAKE_LOCK_PARTIAL | Ti.Android.WAKE_LOCK_ON_AFTER_RELEASE
			}
		});
		Titanium.Android.NotificationManager.notify(1, notification);
	}, 4000);
});

win.open();
```

### CODE REVIEW

File: apidoc/Titanium/Android/Notification.yml [hunk 1]
```diff
@@ -366,6 +366,22 @@ properties:
     since: "3.6.0"
     osver: {android: {min: "5.0"}}
 
+  - name: wakeLock
+    summary: Will wake up the device for the given time (in milliseconds) when the notification is shown.
+        The application needs to also set the `android.permission.WAKE_LOCK` permission
+        in the Android manifest section of the `tiapp.xml` file.
+
```
Comment (hansemannn) [inline apidoc/Titanium/Android/Notification.yml]: Replace with:
```yml
summary: |
    Will wake up the device for the given time (in milliseconds) when the notification is shown.
    The application needs to also set the `android.permission.WAKE_LOCK` permission
    in the Android manifest section of the `tiapp.xml` file.
```

File: android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationManagerModule.java [hunk 2]
```diff
@@ -71,6 +76,18 @@ public void cancelAll()
...
 	{
 		getManager().notify(id, notificationProxy.buildNotification());
+
+		int wakeTime = notificationProxy.wakeTime;
+		if (wakeTime > 0){
+			PowerManager pm = (PowerManager)TiApplication.getInstance().getSystemService(TiApplication.getInstance().getApplicationContext().POWER_SERVICE);
+			boolean isScreenOn = pm.isScreenOn();
+			if(isScreenOn==false) {
+				WakeLock wl = pm.newWakeLock(PowerManager.FULL_WAKE_LOCK | PowerManager.ACQUIRE_CAUSES_WAKEUP | PowerManager.ON_AFTER_RELEASE, "MyWakeLock");
+    			wl.acquire(wakeTime);
```
Comment (hansemannn) [inline android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationManagerModule.java]: Indentation in 86 and space in `){`. Maybe rename `MyWakeLock` to `TiWakeLock`.

**Review by hansemannn — CHANGES_REQUESTED** [review]

File: android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationManagerModule.java [hunk 3]
```diff
@@ -9,11 +9,16 @@
...
+import org.appcelerator.titanium.TiC;
+import org.appcelerator.titanium.util.TiConvert;
 
 import ti.modules.titanium.android.AndroidModule;
 import android.app.Activity;
 import android.app.Notification;
 import android.app.NotificationManager;
+import android.os.PowerManager;
+import android.os.PowerManager.WakeLock;
+
```
Comment (garymathews) [inline android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationManagerModule.java]: Also remove this extra line :)

File: android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationManagerModule.java [hunk 4]
```diff
@@ -71,6 +76,18 @@ public void cancelAll()
 	public void notify(int id, NotificationProxy notificationProxy)
 	{
 		getManager().notify(id, notificationProxy.buildNotification());
+
+		int wakeTime = notificationProxy.wakeTime;
+		if (wakeTime > 0){
+			PowerManager pm = (PowerManager)TiApplication.getInstance().getSystemService(TiApplication.getInstance().getApplicationContext().POWER_SERVICE);
+			boolean isScreenOn = pm.isScreenOn();
+			if(isScreenOn==false) {
```
Comment (hansemannn) [inline android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/NotificationManagerModule.java]: Same here btw: Spaces spaces spaces 👾

**Comment by m1ga** [comment]
Change the parameter to accept a HashMap like this: `wakeLock:{time: 1000, flag: Ti.Android.PARTIAL_WAKE_LOCK}` to be more flexible.

File: apidoc/Titanium/Android/Notification.yml [hunk 5]
```diff
@@ -366,9 +366,44 @@ properties:
...
+    type: int
+    default: 3000
+  - name: flags
+    summary: Wake lock level and flag. See [Android Developers: PowerManager](https://developer.android.com/reference/android/os/PowerManager.html#newWakeLock(int,%20java.lang.String))
+    type: int
+    constants: |
+        Titanium.Android.PARTIAL_WAKE_LOCK | Titanium.Android.FULL_WAKE_LOCK | 
+        Titanium.Android.SCREEN_DIM_WAKE_LOCK | Titanium.Android.SCREEN_BRIGHT_WAKE_LOCK
+        Titanium.Android.ACQUIRE_CAUSES_WAKEUP | Titanium.Android.ON_AFTER_RELEASE
+    default: Titanium.Android.FULL_WAKE_LOCK | Titanium.Android.ACQUIRE_CAUSES_WAKEUP | Titanium.Android.ON_AFTER_RELEASE
```
Comment (hansemannn) [inline apidoc/Titanium/Android/Notification.yml]: We usually prefix the constants so it's easier to find and group them. For example:
- WAKE_LOCK_PARTIAL
- WAKE_LOCK_FULL
- WAKE_LOCK_SCREEN_DIM

Then you could also also write `constants: [Titanium.Android_WAKE_LOCK_*]`

Comment (m1ga) [inline apidoc/Titanium/Android/Notification.yml]: Ah, nice! I've made the changes.

File: apidoc/Titanium/Android/Notification.yml [hunk 6]
```diff
@@ -366,9 +366,41 @@ properties:
...
+since: 6.2.0
+properties:
+  - name: time
+    summary: Minimum time the device will be switched on (plus device switch-off time)
+    type: int
+    default: 3000
+  - name: flags
+    summary: Wake lock level and flag. See [Android Developers: PowerManager](https://developer.android.com/reference/android/os/PowerManager.html#newWakeLock(int,%20java.lang.String))
+    type: int
+    constants: [Titanium.Android_WAKE_LOCK_*]
```
Comment (hansemannn) [inline apidoc/Titanium/Android/Notification.yml]: `Titanium.Android_WAKE_LOCK_*` -> `Titanium.Android.WAKE_LOCK_*`

File: android/modules/android/src/java/ti/modules/titanium/android/AndroidModule.java [hunk 7]
```diff
@@ -272,6 +273,13 @@
 	@Kroll.constant public static final int NAVIGATION_MODE_STANDARD = ActionBar.NAVIGATION_MODE_STANDARD;
 	@Kroll.constant public static final int NAVIGATION_MODE_TABS = ActionBar.NAVIGATION_MODE_TABS;
 
+	@Kroll.constant public static final int WAKE_LOCK_PARTIAL = PowerManager.PARTIAL_WAKE_LOCK;
+	@Kroll.constant public static final int WAKE_LOCK_FULL = PowerManager.FULL_WAKE_LOCK;
+	@Kroll.constant public static final int WAKE_LOCK_SCREEN_DIM = PowerManager.SCREEN_DIM_WAKE_LOCK;
+	@Kroll.constant public static final int WAKE_LOCK_SCREEN_BRIGHT = PowerManager.SCREEN_BRIGHT_WAKE_LOCK;
+	@Kroll.constant public static final int WAKE_LOCK_ACQUIRE_CAUSES_WAKEUP = PowerManager.ACQUIRE_CAUSES_WAKEUP;
+	@Kroll.constant public static final int WAKE_LOCK_ON_AFTER_RELEASE = PowerManager.ON_AFTER_RELEASE;
```
Comment (hansemannn) [inline android/modules/android/src/java/ti/modules/titanium/android/AndroidModule.java]: Missing docs for the constants?

**Review by hansemannn — CHANGES_REQUESTED** [review]

File: android/modules/android/src/java/ti/modules/titanium/android/AndroidModule.java (hunk 7 above)
Comment (m1ga) [inline android/modules/android/src/java/ti/modules/titanium/android/AndroidModule.java]: missing file :-)

**Comment by hansemannn** [comment]
LGMT! @garymathews for an additional review.

**Review by hansemannn — APPROVED** [review]

**Comment by hansemannn** [comment]
@m1ga You closed the PR? :(

**Comment by m1ga** [comment]
oh sorry. closed the wrong branch

**Review by garymathews — APPROVED** [review]
CR: **PASS**

Conflict needs fixing though

**Comment by hansemannn** [comment]
I think my latest resolve of the merge-conflicts screw up around [here](https://github.com/appcelerator/titanium_mobile/pull/8990/commits/2390370b481cffe295491d0c66f96f3756229d53#diff-ca6cb052e5d20d93fc0baecf77560c3bR128). Also, the backport #9307 has a different number of lines and removed the `dragstart` event-key from the code, no idea why (it was only removed in an earlier commit but restored later). So we need to check those issue first before giving it to QE.

**Comment by lokeshchdhry** [comment]
@garymathews , @m1ga , 
I am having issues when I try to test. I am checking with the above code.
It works fine with `flags: Ti.Android.PARTIAL_WAKE_LOCK`, but not with any other flag. The screen does not turn on at full brightness or in dim state when the notification arrives. 
Also, for flag `flags: Ti.Android.WAKE_LOCK_ACQUIRE_CAUSES_WAKEUP` I am getting the below errors:
```
[WARN] :   W/System.err: java.lang.IllegalArgumentException: Must specify a valid wake lock level.
[WARN] :   W/System.err: 	at android.os.PowerManager.validateWakeLockParameters(PowerManager.java:517)
[WARN] :   W/System.err: 	at android.os.PowerManager.newWakeLock(PowerManager.java:501)
[WARN] :   W/System.err: 	at ti.modules.titanium.android.notificationmanager.NotificationManagerModule.notify(NotificationManagerModule.java:89)
[WARN] :   W/System.err: 	at org.appcelerator.kroll.runtime.v8.V8Function.nativeInvoke(Native Method)
[WARN] :   W/System.err: 	at org.appcelerator.kroll.runtime.v8.V8Function.callSync(V8Function.java:57)
[WARN] :   W/System.err: 	at org.appcelerator.kroll.runtime.v8.V8Function.call(V8Function.java:43)
[WARN] :   W/System.err: 	at ti.modules.titanium.TitaniumModule$Timer.run(TitaniumModule.java:147)
[WARN] :   W/System.err: 	at android.os.Handler.handleCallback(Handler.java:739)
[WARN] :   W/System.err: 	at android.os.Handler.dispatchMessage(Handler.java:95)
[WARN] :   W/System.err: 	at android.os.Looper.loop(Looper.java:148)
[WARN] :   W/System.err: 	at android.app.ActivityThread.main(ActivityThread.java:5417)
[WARN] :   W/System.err: 	at java.lang.reflect.Method.invoke(Native Method)
[WARN] :   W/System.err: 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
[WARN] :   W/System.err: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
[ERROR] :  TiExceptionHandler: (main) [4352,4352] ----- Titanium Javascript Runtime Error -----
[ERROR] :  TiExceptionHandler: (main) [0,4352] - In /app.js:31,40
[ERROR] :  TiExceptionHandler: (main) [1,4353] - Message: Uncaught Must specify a valid wake lock level.
[ERROR] :  TiExceptionHandler: (main) [0,4353] - Source: 		Titanium.Android.NotificationManager.notify(1, notification);
[ERROR] :  V8Exception: Exception occurred at /app.js:31: Uncaught Must specify a valid wake lock level.
[ERROR] :  V8Exception: Must specify a valid wake lock level.
```

**Comment by lokeshchdhry** [comment]
Failing FR for now.

**Comment by m1ga** [comment]
ACQUIRE_CAUSES_WAKEUP and ON_AFTER_RELEASE are flags you can add to the WAKE_LOCK levels. See https://developer.android.com/reference/android/os/PowerManager.html
I'll implement a check so it will show an error instead of the crash

**Comment by lokeshchdhry** [comment]
@m1ga , Thanks for the link I was using the flags in a wrong way. It works fine. But with flag `WAKE_LOCK_SCREEN_BRIGHT` & `WAKE_LOCK_FULL` the screen turns `ON` but not at the full brightness.

**Comment by garymathews** [comment]
@lokeshchdhry here's another test case

```JS
var win = Ti.UI.createWindow({
        title: 'TIMOB-24629',
        backgroundColor: 'gray'
    }),
    btn = Ti.UI.createButton({
        title: 'START',
        width: '33%'
    }),
    ntn = Titanium.Android.createNotification({
        contentTitle: 'WAKE-LOCK TEST',
        contentText: 'Do nothing...',
        wakeLock: {
            time: 3000,
            flags: Ti.Android.WAKE_LOCK_FULL | Ti.Android.WAKE_LOCK_ACQUIRE_CAUSES_WAKEUP | Ti.Android.WAKE_LOCK_ON_AFTER_RELEASE
        }
    });

btn.addEventListener('click', function() {
    var i = 4,
        interval = setInterval(function() {
            btn.setTitle('COUNT ' + --i);
            if (i <= 0) {
                Ti.Android.NotificationManager.notify(1, ntn);
                clearInterval(interval);
                btn.setTitle('START');
                btn.setEnabled(true);
            }
        }, 1000);
    btn.setEnabled(false);
    btn.setTitle('COUNT ' + --i);
});

win.add(btn);
win.open();
```

**Comment by lokeshchdhry** [comment]
Thanks @garymathews , I was using the flags in a wrong way. But, as I have commented above with flag `WAKE_LOCK_SCREEN_BRIGHT` & `WAKE_LOCK_FULL` the screen turns `ON` but not at the full brightness. 
Is this expected ?

**Comment by garymathews** [comment]
@lokeshchdhry I would say it's expected behaviour since the flags are being set correctly, that behaviour is dictated by Android.

**Pull request merged**